car/README.md

```diff
@@ -0,0 +1,108 @@
+
+The actual format is just a (mostly) recursively-defined topological sort of the
+desired DAG with some metadata for fast traversal.
+
+```
+CID
+len(ROOT)
+ROOT
+[
+  CHILD-1-OFFSET
+  ...
+]
+[
+  len(CHILD-1)
+  CHILD-1
+  [
+    CHILD-1/1-OFFSET
+    ...
+  ]
+  ...
+]
+```
+
+Offsets are relative, offsets for missing children use a sentinel value.
+
+We only bother including the root CID because all the other CIDs are embedded in
+the objects themselves. This saves space and *forces* parsers to actually
+traverse the DAG (hopefully validating it).
+
+## Motivation
+
+Use cases:
+
+1. Reliably export/import a DAG to/from an external hard drive (backup, sneakernet).
+2. Traverse a large DAG on an external hard drive without parsing the *entire* DAG.
+3. Traverse a large DAG streamed over HTTP without downloading the entire thing.
+
+The simple method is to copy the entire repo. However, for performance, we need
+to be able to upgrade the repo format so this isn't really a stable format.
+Additionally, repos need to support insertions, deletions, and random lookups.
+Supporting these efficiently necessarily complicates the formats. We'd like
+something simple and portable (backup).
+
+The slightly more complex way is to download every object into a separate file
+and then import each file. However, this isn't very convenient and *does not*
+scale to large directories well (use case 2).
+
+One could improve this multi-file approach by splitting up the DAG into multiple
+directories and providing a set of tools to manage the files. However, we'd
+rather not rely on the filesystem for anything, really. Filesystems:
+
+* Don't always deal with names well (e.g., FAT16).
+* Don't always handle many small files well.
+* Aren't usually as space-efficient as possible (to support updates).
+* Are complex (easy to corrupt metadata/structure).
+
+Additionally, it's hard to download a directory structure over HTTP (motivated by
+use-case 3). One can just TAR it up but that layers another (complex) file
+format into the mix.
+
+So, we'd like a new single-file format that, if necessary, we can just `dd` to a
+drive in place of a filesystem.
+
+TODO: Expand.
+
+# Questions
+
+However, there are a few open questions.
+
+## Uint64/Varint
+
+The advantage of using uint64s over varints is that we can leave the "jump
+tables" blank and then fill them in on a second pass after we've written
+everything. However, if we topologically sort the DAG, we may be able to compute
+the jump tables up-front.
+
+The advantages of varints over uint64 are space and flexibility (DAGs larger
+than 16 Exbibytes).
+
+Currently, I'm leaning toward varints as this will make storing lots of small
+blocks significantly more efficient.
+
+## Inline Blocks
+
+So, we can technically have inline blocks using the identity multihash. How do
+we deal with them?
+
+1. We *don't* want to duplicate the data.
+2. We need to support inline blocks with children.
+
+## Topological sort
```

Comment:
I think the purpose of the topological sort and how exactly it is going to sort needs to be fleshed out a bit more.

Reply:
Totally agree. Basically, it allows us to traverse a DAG in one pass. This + offsets makes traversing a DAG on, e.g., tape really fast.

Reply:
So the lower nodes are stored after the higher nodes? Then the offset calculation will be tricky. I don't see how that can work with the varints. With int64 you could just backfill once you know the offsets.

If you store first the leaf nodes, and then the higher nodes and then the root, you always know the offset when you write a node. But then the access pattern is backwards.

Reply:
> So the lower nodes are stored after the higher nodes?

Yes. 

> With int64 you could just backfill once you know the offsets.

We (@whyrusleeping and I) planned on using uint64s at first. However:

1. The best topological sort algorithm I could find (basically, just a DFS) actually does work backwards.
2. If we want to provide an index, we can't make the CAR in one pass anyways (although, if we *don't* do a topological sort, we could dump the data in one pass and then write in the index afterwards).

My current plan is to do one pass to determine the structure of the car and a second pass to write it. Unfortunately, this will require a significant amount of scratch space.

Reply:
So it would be something like
1. do the sorting so that leaves are last, create a sequence of nodes
2. go backwards through the seq and determine serialised size for each node (to do this you need the offsets)
3. go forwards through the seq and write to disk

Sounds good. Backfilling the offsets in case of int64 offsets would also need to be done in a clever way if you want linear access patterns. The alternative would be to write leaves first and calculate the offsets in one pass. That would be a single pass for writing, but would be a backwards access pattern for reading.

Reply:
Camlistore, with care taken for data archeology/recoverability, stores data and metadata like folder structures and references the same way: content adressed blobs.

I understand CAR as IPFS's version of tar, to archive IPFS-held data in a (tape) streaming friendly format. Maybe it could become a .torrent competitor. But maybe that functionality can be achieved in a simpler way? Just pushing blobs to/from storage automatically, where automation can be made from a combination of Bloom filters and Inverted Bloom filters over the content hashes. This is currently being implemented for bitcoin block propagation, based on fresh insights, that surely can be recycled also for IPFS.

Imagine this upgrade: A HTTP/2 server can stream with interleaving so that several "files" and "directory structures" could be handled at once.

Which supports this use case: you'd click a web link into your "junior year reception party" and your HTTP/2 server suggest you might also want the gag/for fun subtexts or voice over that other people who streamed that video also used. Your client, say VLC, makes it a single keypress to say yes to any of the suggested extras.